File: src/bmsx/vm/ide/editor_undo.ts
import type { PieceTreeNode } from './piece_tree_buffer';

export type TextUndoKind = 'insert' | 'delete' | 'replace';

export class TextUndoOp {
	public kind: TextUndoKind = 'insert';
	public offset = 0;

	public deletedLen = 0;
	public insertedLen = 0;

	public deletedRoot: PieceTreeNode | null = null;
	public insertedRoot: PieceTreeNode | null = null;

	public setInsert(offset: number, insertedLen: number): void {
		this.kind = 'insert';
		this.offset = offset;
		this.deletedLen = 0;
		this.insertedLen = insertedLen;
		this.deletedRoot = null;
		this.insertedRoot = null;
	}

	public setDelete(offset: number, deletedLen: number, deletedRoot: PieceTreeNode | null): void {
		this.kind = 'delete';
		this.offset = offset;
		this.deletedLen = deletedLen;
		this.insertedLen = 0;
		this.deletedRoot = deletedRoot;
		this.insertedRoot = null;
	}

	public setReplace(offset: number, deletedLen: number, deletedRoot: PieceTreeNode | null, insertedLen: number): void {
		this.kind = 'replace';
		this.offset = offset;
		this.deletedLen = deletedLen;
		this.insertedLen = insertedLen;
		this.deletedRoot = deletedRoot;
		this.insertedRoot = null;
	}
}

export class EditorUndoRecord {
	public readonly ops: TextUndoOp[] = [];

	public beforeCursorRow = 0;
	public beforeCursorColumn = 0;
	public beforeScrollRow = 0;
	public beforeScrollColumn = 0;
	public beforeHasSelectionAnchor = false;
	public beforeSelectionAnchorRow = 0;
	public beforeSelectionAnchorColumn = 0;

	public afterCursorRow = 0;
	public afterCursorColumn = 0;
	public afterScrollRow = 0;
	public afterScrollColumn = 0;
	public afterHasSelectionAnchor = false;
	public afterSelectionAnchorRow = 0;
	public afterSelectionAnchorColumn = 0;

	public setBeforeState(
		cursorRow: number,
		cursorColumn: number,
		scrollRow: number,
		scrollColumn: number,
		selectionAnchorRow: number,
		selectionAnchorColumn: number,
		hasSelectionAnchor: boolean,
	): void {
		this.beforeCursorRow = cursorRow;
		this.beforeCursorColumn = cursorColumn;
		this.beforeScrollRow = scrollRow;
		this.beforeScrollColumn = scrollColumn;
		this.beforeHasSelectionAnchor = hasSelectionAnchor;
		this.beforeSelectionAnchorRow = selectionAnchorRow;
		this.beforeSelectionAnchorColumn = selectionAnchorColumn;
	}

	public setAfterState(
		cursorRow: number,
		cursorColumn: number,
		scrollRow: number,
		scrollColumn: number,
		selectionAnchorRow: number,
		selectionAnchorColumn: number,
		hasSelectionAnchor: boolean,
	): void {
		this.afterCursorRow = cursorRow;
		this.afterCursorColumn = cursorColumn;
		this.afterScrollRow = scrollRow;
		this.afterScrollColumn = scrollColumn;
		this.afterHasSelectionAnchor = hasSelectionAnchor;
		this.afterSelectionAnchorRow = selectionAnchorRow;
		this.afterSelectionAnchorColumn = selectionAnchorColumn;
	}
}

File: packages/bmsx-console/src/res/_ignore/ide/editor_undo.lua
-- undo.lua

local undo = {}

local text_undo_op = {}
text_undo_op.__index = text_undo_op

function text_undo_op.new()
	return setmetatable({
		kind = "insert",
		offset = 0,
		deleted_len = 0,
		inserted_len = 0,
		deleted_root = nil,
		inserted_root = nil,
	}, text_undo_op)
end

function text_undo_op:set_insert(offset, inserted_len)
	self.kind = "insert"
	self.offset = offset
	self.deleted_len = 0
	self.inserted_len = inserted_len
	self.deleted_root = nil
	self.inserted_root = nil
end

function text_undo_op:set_delete(offset, deleted_len, deleted_root)
	self.kind = "delete"
	self.offset = offset
	self.deleted_len = deleted_len
	self.inserted_len = 0
	self.deleted_root = deleted_root
	self.inserted_root = nil
end

function text_undo_op:set_replace(offset, deleted_len, deleted_root, inserted_len)
	self.kind = "replace"
	self.offset = offset
	self.deleted_len = deleted_len
	self.inserted_len = inserted_len
	self.deleted_root = deleted_root
	self.inserted_root = nil
end

local undo_record = {}
undo_record.__index = undo_record

function undo_record.new()
	return setmetatable({
		ops = {},
		before_cursor_row = 0,
		before_cursor_column = 0,
		before_scroll_row = 0,
		before_scroll_column = 0,
		before_has_selection_anchor = false,
		before_selection_anchor_row = 0,
		before_selection_anchor_column = 0,
		after_cursor_row = 0,
		after_cursor_column = 0,
		after_scroll_row = 0,
		after_scroll_column = 0,
		after_has_selection_anchor = false,
		after_selection_anchor_row = 0,
		after_selection_anchor_column = 0,
	}, undo_record)
end

function undo_record:set_before_state(cursor_row, cursor_column, scroll_row, scroll_column, selection_anchor_row, selection_anchor_column, has_selection_anchor)
	self.before_cursor_row = cursor_row
	self.before_cursor_column = cursor_column
	self.before_scroll_row = scroll_row
	self.before_scroll_column = scroll_column
	self.before_has_selection_anchor = has_selection_anchor
	self.before_selection_anchor_row = selection_anchor_row
	self.before_selection_anchor_column = selection_anchor_column
end

function undo_record:set_after_state(cursor_row, cursor_column, scroll_row, scroll_column, selection_anchor_row, selection_anchor_column, has_selection_anchor)
	self.after_cursor_row = cursor_row
	self.after_cursor_column = cursor_column
	self.after_scroll_row = scroll_row
	self.after_scroll_column = scroll_column
	self.after_has_selection_anchor = has_selection_anchor
	self.after_selection_anchor_row = selection_anchor_row
	self.after_selection_anchor_column = selection_anchor_column
end

undo.text_undo_op = text_undo_op
undo.undo_record = undo_record

return undo
